[FEATURE] Remplir et désactiver les champs nom et prénom de l'utilisateur connecté via le GAR lors de la réconciliation (PIX-1118).

yahyayouness opened the pull request:
## :unicorn: Problème

Aujourd'hui, lorsque qu'une requête provenant du GAR est reçue, un utilisateur est créé grâce aux informations contenues dans la requête (prénom, nom et samlId). 
Dans le cadre de la pérennité des comptes, on veut empêcher la création des comptes avant que l'élève ne se soit réconcilié avec son établissement scolaire.

## :robot: Solution
Suite au ticket https://1024pix.atlassian.net/browse/PIX-1039, les informations permettant de créer l'utilisateur sont présent dans un token id disponible depuis l'url (en tant que queryParam). Lors de l'accès au formulaire de réconciliation, les champs prénom et nom sont pré-remplis avec les informations de l'utilisateur. Etant donné que le compte utilisateur ne sera, à terme, plus créé il faut utiliser les informations contenues dans le token afin de pré-remplir ces champs.

La solution consiste donc à:

- décoder le token pour extraire le nom/prénom
- afficher le nom/prénom dans le formulaire (non éditable)
- Le renvoi des données au back (token) sera fait uniquement lors de la réconciliation (PIX-1119)


## :rainbow: Remarques
Un helper jwt a été crée pour décoder le token.

## :100: Pour tester

Cliquer sur ce [lien](https://app-pr1793.review.pix.fr/campagnes?externalUser=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmaXJzdF9uYW1lIjoiRmlyc3QiLCJsYXN0X25hbWUiOiJMYXN0Iiwic2FtbF9pZCI6InNhbWxJRHFzZnNmcWZxZnNxZmhmZmdyciIsImlhdCI6MTU5NzkyOTQ0OCwiZXhwIjoxNTk3OTMzMDQ4fQ.KRh6ZKr6EwM1QvveTHsWush6z9meVAI6enVYgSQ-MuI)
Vérifier que le localStorage contient bien un attribut externalUsercontenant le token dans la clé ember_simple_auth-session.
Rentrer un code parcours = RESTRICTD et vérifier que l'on arrive bien sur le formulaire de réconciliation.
Vérifier que les deux champs prénom et nom sont désactivé et pré-remplis avec les valeurs de l'utilisateur externe.

Commit:
[API] Add helper to decode jwt token

Commit:
[API] Decode and fill names Inputs if ExternalUser

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr1793.review.pix.fr
- Orga: https://orga-pr1793.review.pix.fr
- Certif: https://certif-pr1793.review.pix.fr
- Admin: https://admin-pr1793.review.pix.fr
- API: https://api-pr1793.review.pix.fr/api/

 Please check it out!

Commit:
[API] remove login with external source and refactor test

Commit:
[E2E] Evolve feature join campaign from GAR

We cannot test assertion of e2e test until the feature is complete. This feature is evolved to fill the inputs username,lastname if the user is autenticated from GAR and will be fixed as soon the feature is done.

octo-topi commented:
Il manque un point au bout du titre de la PR (je ne le corrige pas car je sais que tu souhaites le faire toi même) :grinning:

MathieuGilet commented on high-level-tests/e2e/cypress/integration/pix-app/campaign-assessment.feature:
```diff
@@ -51,13 +51,12 @@ Fonctionnalité: Campagne d'évaluation
-    Alors je vois la page de "presentation" de la campagne
-    Lorsque je clique sur "Je commence"
-    Alors je vois la page de "didacticiel" de la campagne
+#    Alors je vois la page de "presentation" de la campagne
```
Le scénarios sont commentés ici car il ne peuvent pas être pleinement réalisés dans ce ticket mais le seront grâce au ticket PIX-1119 (la PR sera ouverte très prochainement).

yahyayouness replied:
Yes, j'avais mis aussi le détail dans la description du commit.

MathieuGilet commented on mon-pix/app/components/routes/campaigns/restricted/join-sco.js:
```diff
@@ -28,6 +29,9 @@ class Validation {
 }
 
 export default class JoinSco extends Component {
+  queryParams = ['participantExternalId'];
+  participantExternalId = null;
+
```
Pas utile de rajouter ces deux lignes dans un composant, elle n'ont aucun effet.

yahyayouness replied:
C'est suite au rebase qu'elles ont été rajoutés, on peut faire une passe pour voir si la résolution du conflit s'est bien déroulée.

yahyayouness replied:
Done

MathieuGilet commented on mon-pix/app/components/routes/campaigns/restricted/join-sco.js:
```diff
@@ -47,13 +51,16 @@ export default class JoinSco extends Component {
   @tracked dayOfBirth = '';
   @tracked monthOfBirth = '';
   @tracked yearOfBirth = '';
+  @tracked studentNumber = '';
```
Le `studentNumber` n'est pas nécessaire ici, il a été supprimé dans une précédente PR.

yahyayouness replied:
Ok thanks, je recheck le rebase, pour mettre à niveau ce fichier.

yahyayouness replied:
Done

MathieuGilet commented on mon-pix/app/components/routes/campaigns/restricted/join-sco.js:
```diff
@@ -70,7 +77,7 @@ export default class JoinSco extends Component {
 
   get isDisabled() {
-    return this.session.data.authenticated.source === 'external';
+    return (undefined !== get(this.session,'data.externalUser'));
```
```suggestion
    return get(this.session,'data.externalUser');
```

thloesch commented on mon-pix/app/helpers/jwt.js:
```diff
@@ -0,0 +1,10 @@
+import { helper } from '@ember/component/helper';
+
+export function decodeToken(accessToken)
```
Si on crée un nouveau helper pour l'occasion, je pense qu'il faudrait l'utiliser où il peut l'être.
Je pense à `get-authenticated-user.js` et `oauth2.js`

yahyayouness replied:
Done, c'était prévu, j'ai oublié de le faire, thanks.

Commit:
[IHM] replace extractDataFromToken by decode token helper

Commit:
[IHM] Use decode token helper in getAuthenticatedUser
